### PR TITLE
[don't merge] this PR is for dev code comparisons/discussions

### PR DESCRIPTION
**Pull request by wyli**
for the convenience of comparing `dev` vs `feature/MetaTensor`

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Minor comments:
1. Is it better to use `torch.empty()` here?
https://github.com/Project-MONAI/MONAI/blob/feature/MetaTensor/monai/utils/type_conversion.py#L54
2. I see the `convert_to_tensor()` and `convert_to_tensor()` are very similar, maybe we can combine them or at least simplify them?

Thanks.

**Comment by wyli** [comment]
some design decisions we had today:

- each transform and the inverse will only change the most recent element in `applied_operations`
- each transform's input can be numpy/torch/metatensor
- each transform's output depends on `get_track_meta`, flag on: output metatensor with default meta if needed, flag off: output torch tensor cc @Project-MONAI/core-reviewers

File: monai/data/meta_tensor.py [hunk 1]
```diff
@@ -181,30 +184,34 @@ def update_meta(rets: Sequence, func, args, kwargs) -> Sequence:
                 ret = ret.as_tensor()
             # else, handle the `MetaTensor` metadata.
             else:
-                meta_args = MetaObj.flatten_meta_objs(list(args) + list(kwargs.values()))
-                ret._copy_meta(meta_args)
+                meta_args = MetaObj.flatten_meta_objs(args, kwargs.values())  # type: ignore
+                ret._copy_meta(meta_args, deep_copy=not is_batch)
```
Comment (wyli) [inline monai/data/meta_tensor.py:188]: this is a workaround for not deepcopying attributes when `metatensor.is_batch` flag is `True`

File: monai/data/meta_tensor.py [hunk 2]
```diff
@@ -125,19 +126,20 @@ def __init__(
...
-            setattr(self, attribute, val.to(self.device))
+    def _copy_attr(self, attributes: list[str], input_objs, defaults: list, deep_copy: bool) -> None:
+        super()._copy_attr(attributes, input_objs, defaults, deep_copy)
+        for a in attributes:
+            val = getattr(self, a)
+            if isinstance(val, torch.Tensor):
+                setattr(self, a, val.to(self.device))
 
     @staticmethod
     def update_meta(rets: Sequence, func, args, kwargs) -> Sequence:
```
Comment (wyli) [inline monai/data/meta_tensor.py:145]: may have "strict" and "dev" modes to control manual modifications of the meta information

File: monai/data/meta_tensor.py [hunk 3]
```diff
@@ -125,19 +126,20 @@ def __init__(
         elif isinstance(x, MetaTensor):
             self.applied_operations = x.applied_operations
         else:
-            self.applied_operations = self.get_default_applied_operations()
+            self.applied_operations = MetaObj.get_default_applied_operations()
```
Comment (wyli) [inline monai/data/meta_tensor.py:129]: should revisit the docstrings of `MetaTensor` and `MetaObj`.

File: tests/test_crop_base.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    #     for param_type in TEST_NDARRAYS + (None,):
+    #         with self.subTest(param_type=param_type):
+    #             roi_params_mod = deepcopy(roi_params)
+    #             if param_type is not None:
+    #                 for k in ("roi_start", "roi_end", "roi_center", "roi_size"):
+    #                     if k in roi_params:
+    #                         roi_params_mod[k] = param_type(roi_params[k])
+    #     slices = CropBase.calculate_slices(**roi_params)
+    #     self.assertEqual(slices, expected_slices)
+
```
Comment (wyli) [inline tests/test_crop_base.py:77]: hi @rijobro, do you know some context of this test case? is it to be implemented or deprecated? I forget the reason of having this block...

File: monai/transforms/croppad/dictionary.py [hunk 5]
```diff
@@ -298,37 +290,60 @@ def __init__(
...
+        if isinstance(self.cropper, Randomizable):
+            self.cropper.set_random_state(seed, state)
+        return self
+
+    def randomize(self, img_size: Sequence[int]) -> None:
+        if isinstance(self.cropper, Randomizable):
+            self.cropper.randomize(img_size)
+
+    def __call__(self, data: Mapping[Hashable, torch.Tensor]) -> Dict[Hashable, torch.Tensor]:
+        d = dict(data)
```
[the remark this answers is not in the document]
Comment (wyli) [reply monai/transforms/croppad/dictionary.py:332]: it's needed because the dictionary size shouldn't change in a loop

**Pull request closed without merging**